### DISCUSSION TITLE
passing values between visitor calls

### DISCUSSION THREAD

**Discussion by udif** [MikePopoloski/slang#953] [Q&A]
Giving a code like this:
```
    auto instances = compilation->getRoot().topInstances;
    for (auto& i : instances)
        i->visit(makeVisitor([&](auto& visitor, const InstanceSymbol& type) {
...
            // send an argument to visitDefault()
            visitor.visitDefault(type);
...
            // got something back from visitDefault()
            }
        }));
```

How can I pass values between recursive visitor calls? I think I would like to pass arguments from upper parts of the tree sections, as well as returning a return values from lower level parts of the tree back to the upper sections
Let's say I want to recursively visit an `always` block and collect a list of all identifiers to check if all of them are in the sensitivity list.
Do I have to traverse the AST Tree manually, or can I use `makeVisitor()`?

**Comment by MikePopoloski** [comment]
You might be able to make it work by storing some state in the lambda but it's probably more straightforward to just make your own subclass of ASTVisitor and then you can store whatever state you need in plain old variables.

**Comment by udif** [comment]
I already had something working using static strings inside the lambda function (as in the commented out code [here](https://github.com/MikePopoloski/slang/pull/934/commits/28783e8290124e5c5fa683f3c88b9dbfe7f47699), but I was hoping for something more elegant.

**Reply by yanggeorge** [reply]
So far as I know, it is common in visitor design pattern, which visit() method will not return value. All states are kept in visitor.
